### ISSUE TITLE
Schematic file format for 1.13 and newer is not supported

### ISSUE THREAD

**Issue by erich666** [erich666/Mineways#40]
Evidently FAWE supports output of a new schematic file format for their system for Minecraft: https://www.spigotmc.org/resources/fast-async-worldedit-voxelsniper.13932/

Unfortunately, this format is not documented in any place I could find. The traditional, pre-1.13 schematic format is documented here: https://minecraft.gamepedia.com/Schematic_file_format

That said, I suspect someone could go through the [code starting here in FAWE](https://github.com/boy0001/FastAsyncWorldedit/blob/master/core/src/main/java/com/sk89q/worldedit/extent/clipboard/io/) and figure out what they do.

The new and old are considerably different, since 1.13 and pre-1.13 are considerably different. The pre-1.13 format looks like this:
![image](https://user-images.githubusercontent.com/803995/52542584-39737b00-2d6f-11e9-9b75-c5ca228ebe29.png)

It's pretty simple: XYZ dimensions, then the block and data fields in huge arrays.

The new format is more involved:
![image](https://user-images.githubusercontent.com/803995/52542601-77709f00-2d6f-11e9-8089-1d275d9f5b46.png)

It doesn't look all that bad, until you poke around.

There's some Metadata - not sure about what that is, but probably easy to figure out. The Palette is also fairly obvious, but with a huge number of options to parse:
![image](https://user-images.githubusercontent.com/803995/52542645-04b3f380-2d70-11e9-942d-9febf019f188.png)

The TileEntities also look "interesting":

![image](https://user-images.githubusercontent.com/803995/52542660-257c4900-2d70-11e9-9bbe-52988f6d279a.png)

BlockData is confusing. It is an array of size 30,518,469
But the dimensions Height * Length * Width, 230 * 436 * 304, multiply to 30,485,120 - a difference of 33,349. Why?

Basically, this looks like a lot of work to read in, having to parse and convert all possible Palette entries is a fair bit of mindless work. So, I'm passing for now, as there are more interesting things in the world to do than laborious palette parsing for a mod's feature. However, I will add code to see if a schematic format being read in is this new format and, if so, pop up a warning.

**Comment by erich666** [comment]
Added code to issue this error on read:
![image](https://user-images.githubusercontent.com/803995/52543157-4c894980-2d75-11e9-9dbe-9628821f119a.png)

I also now issue a warning if you try to export a 1.13 world and a 1.13 block type is encountered:

![image](https://user-images.githubusercontent.com/803995/52543439-b73b8480-2d77-11e9-9179-c11df42e3fb6.png)

**Comment by IronApollo** [comment]
Hello. I'm currently the lead (and in all honesty, the only) developer of FAWE. Things have been slow and I'm tied up IRL, but I'm here.

The schematic format supported by FAWE and WE prior to 1.13 was the MCEDIT schematic format. Due to the changes in the way Minecraft handles, saves and reads blocks starting in 1.13, WE and FAWE have moved on from the MCEDIT schematic format to the Sponge schematic format. This format is more suitable and ideal for the 1.13 environment, and hopefully for updates to come. This is a change that likely will not be reverted, and as for FAWE, we no longer support any writing feature related to the MCEDIT schematic format, but reading is still possible.

If you are interested in knowing more about the new default schematic format, you can Google "Sponge schematic format". I would provide a link, but I'm mobile at the moment.

If you have any further questions regarding FAWE and the way it handles schematics, feel free to message me on Discord @IronApollo#1953, or join the IntellectualSites Discord server to engage in conversation on the topic.

**Comment by erich666** [comment]
Josh, thanks for your reply, and I'm impressed you found this issue at all. Like I say, adding new schematic read support for 1.13 is low on my list; so far only one user has reported needing it. But, I'm happy to know where to look for more information: https://github.com/SpongePowered/Schematic-Specification/blob/master/versions/schematic-1.md

I'll contact you on Discord - I'm definitely interested in other elements of FAWE, such as being able to read "non-classic" Minecraft files formats the way MCEdit does.

**Comment by erich666** [comment]
@IronApollo Hmmm, unfortunately, my attempts to find you or IntellectualSites on Discord have failed. I'd certainly appreciate it if you contact me directly, [I'm here](http://www.realtimerendering.com/erich/minecraft/public/mineways/contact.html). It'd be nice to talk with someone else supporting reading Minecraft files; I promise I won't spam you.

**Comment by craftyfoxe** [comment]
@erich666 Hi, I'm actually the one who asked IronApollo to help with you since you were having a problem and I think if Mineways supported 1.13 schematics that would be awesome to render stuff. 

You can come to the Intellectual Discord server by going to this link: [https://discord.gg/ebreupz](https://discord.gg/ebreupz)

**Comment by acs** [comment]
Hi all,

I have added Schematic write and write support for https://github.com/juntosdesdecasa/mcthings. But using the legacy format for MCEdit, because the amount of creations in this format is larger and because I am working with the Raspberry Pi API (for Python programming) which uses the old block ids definition before the Minecraft flattening: https://minecraft.gamepedia.com/Java_Edition_1.13/Flattening.

I am pretty sure @erich666 that to be future proof, Mineways will need to support the new way to define the blocks in Minecraft, so the move to the support for the Sponge schematic format seems to be a must to use Mineways with all the new Schematics creations (and as time goes by, there will be more Schematics in this new format).

@erich666 thanks for your great tool which is the perfect bridge between Minecraft and the 3D rendering and printing world!

Just my 2 cents!

**Comment by erich666** [comment]
Thanks for your kind words, they mean a lot to me. Unfortunately, updating this feature is a ton of really boring work. But, I'll think about it - boring might be just the thing to keep my mind off the rest of the world. I might figure out some good way to leverage the code that's there.

Duplicate report, see https://github.com/erich666/Mineways/issues/40, so I'm closing it. **But**, thanks for writing, as it led me to the Sponge Format documentation (I didn't know that was what it was called until today).

**Comment by erich666** [comment]
Oh, wups, you put this info in the right issue! Reopened. The Sponge specification is here: https://github.com/SpongePowered/Schematic-Specification

**Comment by acs** [comment]
@erich666 the conversion between blocks ids and materials is already done in some places (like for example, Spigot adds a Legacy Material layer if it finds old plugins loaded that use the old block ids to specify blocks). I have research it a bit in https://github.com/juntosdesdecasa/mcthings/issues/58 and this Legacy support is included in Craftbukkit:

https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/src/main/java/org/bukkit/craftbukkit/legacy/CraftLegacy.java

But it seems it depends in Java libraries from the vanilla Minecraft server:

https://github.com/Akarin-project/Minecraft/blob/master/1.13.2/spigot/net/minecraft/server/DataConverterFlattenData.java

**Comment by erich666** [comment]
@IronApollo @craftyfoxe @acs I could use your help. Do any of you know of a utility that will read in the new-fangled schematic format you've made and convert it/insert it into a Java (Classic) Minecraft world? The utility needs to _not_ be a mod, like WorldEdit.

I have a user of Mineways that wants to read in a schematic. My normal advice is "read it into a world," but they say "the schematic file has flowers floating, if you put it in the world it will be damaged", so the WorldEdit mod can't be used. So, is there another application they could use?

**Comment by IronApollo** [comment]
> @IronApollo @craftyfoxe @acs I could use your help. Do any of you know of a utility that will read in the new-fangled schematic format you've made and convert it/insert it into a Java (Classic) Minecraft world? The utility needs to _not_ be a mod, like WorldEdit.
> 
> I have a user of Mineways that wants to read in a schematic. My normal advice is "read it into a world," but they say "the schematic file has flowers floating, if you put it in the world it will be damaged", so the WorldEdit mod can't be used. So, is there another application they could use?

Hey erich, 

I'm not sure which schematic format you're referring to (I assume .schem aka Sponge Schematic Format). If it is the Sponge Schematic Format you're referring to, it has become the new standard in the world modification scene and I would highly recommend implementing proper support if you find the time!

It's been a while since I have worked on FAWE, but at the time I stopped working on it, this would be a great case to use it as FAWE did not impose block updates on edits as a way to increase world modification speeds. You mention not using a mod, so I don't have a solution for you unless they are willing to paste it into the world using FAWE and see if it keeps their floating flowers. There are important distinctions to be made between FAWE and WE with a lack of block updates in FAWE being one of the largest.

Hope this helps.

**Comment by erich666** [comment]
@IronApollo yes, I meant your new Sponge Schematic Format. And, thanks, I will point them at FAWE.

**Comment by erich666** [comment]
One other way to manipulate .schem files is to use Litematica, [recommended by a person on Reddit](https://www.reddit.com/r/3Dprinting/comments/13u0boy/comment/jmyithc/?%2524deep_link=true&correlation_id=693d441e-33a7-45ae-b4d9-a5f9685083de&ref=email_comment_reply&ref_campaign=email_comment_reply&ref_source=email&%25243p=e_as&_branch_match_id=1155250764236547665&utm_medium=Email%20Amazon%20SES&_branch_referrer=H4sIAAAAAAAAA31O7WrDMAx8muxfkjZ2snZQxqDsNYQbq4k6f6E4mLz9FNb%2BHUhwOt1xN%2Beclo%2B2ZbSWcmNSahyFn1alz6rTKl0QzPImMDJNFIyDld1l3l2V%2Bqq6b5lSSvP0j9ELwbLqmphCpjDJIbTHkBeBR7UebnET9PqD38CPUCI7C4XyDJ4CFrPt8offhBn3ICVZ2iIm2BtW6pp5xaobxsiMzmSKAcgKP5yV1fqItVLmvda9wfqm7bk2%2Ff08nPrDSdndx3gXMXpDDp4FgTG57e8Ho%2FHJ0BT%2BFS1x5RFfkl8jYt2nTgEAAA%3D%3D).

**Comment by erich666** [comment]
@IronApollo @craftyfoxe @acs Chatting with Claude Code on this topic, it points out a few things for writing out this format.

1. v2 only (Recommended)
     Sponge Schematic v2 is what WorldEdit, FAWE, Litematica, and Axiom actually consume. Simpler to write (varint-packed
     BlockData, flat root compound, 2D biomes). Pragmatic baseline.
  2. v3 only
     Newest spec (3D biomes, nested Schematic compound, BlockPalette). Less consumer support today but forward-looking.
  3. Both v2 and v3
     Add a UI selector and emit either version. ~50% more code, more test surface, but covers everyone.

I chose "1" but am open to suggestions. I'm starting with just the writer (Mineways export) for now. For reading, I'll assume "both" makes the most sense.

It then asked:

> 1. Base name only (Recommended)
     Emit just `minecraft:oak_log`, `minecraft:furnace`, etc. Reuse BlockTranslations[] in reverse to look up the canonical
     name from (type, dataVal). Loses orientation/state info but writes correct, consumable schematics for almost any tool.
     ~200 LOC of reverse-lookup work.
  2. Base name + a few common props
     Cover a curated set: axis (logs/pillars), facing (stairs/furnace/ladders), color (wool/concrete/glass), half
     (slabs/stairs), waterlogged. Significant extra work — needs a per-block-family property emitter — but produces materially
     better paste-back results.
  3. Full property fidelity
     Emit every property Minecraft tracks. Effectively writing the inverse of nbt.cpp's readPalette and all per-block translate
      logic. Largest effort and biggest test surface; arguably out of scope for a first pass.

I think "3" is the way to go (kinda "what's the point?" otherwise).

**Comment by erich666** [comment]
Beginnings of a schematic writer are now in branch https://github.com/erich666/Mineways/tree/write_schem

**Comment by erich666** [comment]
Bennthic on the Mineways Discord would like to read .schem files. They write:

for context i was going to make a render for modded world and has its own dimension, so what i did is replace all the modded blocks with vanilla ones using world edit, saved it but unfortuantely can't paste it into the overworld because its way too big of a selection making minecraft run out of ram. So a possible solution was using the already saved .schematic from worldedit and unfortunately it ain't supported now lol

**Comment by erich666** [comment]
Me and Claude Code are trying to add support, starting with export. It's hard. I've made a branch in https://github.com/erich666/Mineways/tree/write_schem with its first attempt.

There's a beta at https://www.realtimerendering.com/erich/minecraft/public/mineways/downloads/Mineways12.13_schem_beta1.zip

It actually works for very simple .schem exports, but fails for more different types of blocks - see below. Without a way to debug into the reader's code further, it's just about impossible for me to add export of .schem files.

Is there some good, simple way to debug reading in .schem files and seeing what went wrong?

[14:15:49] [WorldEdit Task Executor - 0/ERROR]:Uncaught exception occurred in task: Loading schematic test_world.schem
java.io.IOException: Block palette size does not match expected size.
	at knot//com.sk89q.worldedit.extent.clipboard.io.sponge.SpongeSchematicV1Reader.readVersion1(SpongeSchematicV1Reader.java:117)
	at knot//com.sk89q.worldedit.extent.clipboard.io.sponge.SpongeSchematicV2Reader.doRead(SpongeSchematicV2Reader.java:75)
	at knot//com.sk89q.worldedit.extent.clipboard.io.sponge.SpongeSchematicV2Reader.read(SpongeSchematicV2Reader.java:63)
	at knot//com.sk89q.worldedit.command.SchematicCommands$SchematicLoadTask.call(SchematicCommands.java:359)
	at knot//com.sk89q.worldedit.command.SchematicCommands$SchematicLoadTask.call(SchematicCommands.java:341)
	at knot//com.sk89q.worldedit.command.util.AsyncCommandBuilder.runTask(AsyncCommandBuilder.java:174)
	at knot//com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:128)
	at knot//com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:74)
	at knot//com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:80)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
[14:15:49] [Render thread/INFO]:[System] [CHAT] Failed to load schematic: Unknown error. Please see console.

**Comment by erich666** [comment]
The next version of Mineways should have these two features: .schem file import and export, using the modern Sponge schematic file format. Download the beta from https://www.realtimerendering.com/erich/minecraft/public/mineways/downloads/Mineways12.13_schem_beta2.zip and please do let me know of any problems. The file https://github.com/erich666/Mineways/blob/master/docs/mineways.html describes in the Block Types Supported section what (mostly non-graphical) block properties are not exported or imported via .schem.

**Comment by erich666** [comment]
Mineways version 13.00 is out, with import and export of .schem files. Let me know of any bugs.
